lib/pacemaker/pcmk_status.c

```diff
@@ -122,15 +122,20 @@ pacemakerd_status(pcmk__output_t *out)
     rc = pcmk_pacemakerd_api_ping(pacemakerd_api, crm_system_name);
 
     if (rc != pcmk_rc_ok) {
-        /* Got some error from pcmk_pacemakerd_api_ping, so return it. */
-    } else if (state == pcmk_pacemakerd_state_running) {
-        rc = pcmk_rc_ok;
-    } else if (state == pcmk_pacemakerd_state_shutting_down) {
-        rc = ENOTCONN;
-    } else {
-        rc = EAGAIN;
+        // Got some error from pcmk_pacemakerd_api_ping, so return it
+        pcmk_free_ipc_api(pacemakerd_api);
+        return rc;
     }
 
+    switch (state) {
+        case pcmk_pacemakerd_state_running:
+        case pcmk_pacemakerd_state_shutting_down:
+            // CIB may still be accessible while shutting down
+            break;
+        default:
+            rc = ENOTCONN;
+            break;
+    }
```

Comment:
EAGAIN would make sense for the init/starting_daemons/wait_for_ping states, where the cluster is coming up and the caller should try again in a bit. ENOTCONN makes sense for shutdown_complete (and invalid, which we shouldn't see).

However one_shot() would still treat EAGAIN as an error. My gut feeling is that we should report the state and return success, since we were able to get a definitive status.

Reply:
I'm working on something in that vein but I don't really like how it's turning out (aside from the added complexity). I think `crm_mon --one-shot` really *should* return an error if it failed to get the status.

The issue at hand is that it's not even trying to get the status if `pacemakerd` is shutting down.

Reply:
If we stick with the simple approach here, we can still change the return code to `EAGAIN` for the states you mentioned. I agree it makes sense for those.

Reply:
I've got mixed feelings. Getting a response from pacemakerd is successfully getting the status, even if we don't have full CIB/fencing information (yet). I tend to think only failing to communicate with the daemons should be an error exit.

Reply:
See #2905 for an approach that returns success from one-shot mode if we're able to get the `pacemakerd` status. However, as mentioned above, I think it will be fine to return an (appropriate) error code if we get the `pacemakerd` status successfully and still hit an error in `cib_connect()` for example. I don't think it's necessary to return "OK" if we failed to get the cluster status.

Reply:
Hmm yeah, I see your point of view on that. We could just return success from `pcmk__status() no matter what then, if we make it past the `pacemakerd_status()` section. And we could print the `pacemakerd` status from within `pcmk__status()` instead of from within the caller, if we're considering that to be "good enough" as a cluster state result.

Reply:
I think pacemakerd status should be "good enough" unless we're in running state (because then we *should* have CIB and fencing).